gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/tasks/ResolveProductDependenciesTask.java

```diff
@@ -0,0 +1,347 @@
+    }
+
+    /**
+     * A lazy collection of tasks that ensure the {@link Jar} task of any project dependencies from
+     * {@link #productDependenciesConfig} is correctly populated with the recommended product dependencies of that
+     * project, if any (specifically, if they apply the {@link RecommendedProductDependenciesPlugin}).
+     */
+    private Provider<FileCollection> otherProjectProductDependenciesTasks() {
+        return productDependenciesConfig.map(productDeps -> {
+            // Using a ConfigurableFileCollection simply because it implements Buildable and provides a convenient API
+            // to wire up task dependencies to it in a lazy way.
+            ConfigurableFileCollection emptyFileCollection = getProject().files();
+            productDeps.getIncoming().getArtifacts().getArtifacts().stream()
+                    .flatMap(artifact -> {
+                        ComponentIdentifier id = artifact.getId().getComponentIdentifier();
+
+                        // Depend on the ConfigureProductDependenciesTask, if it exists, which will wire up the jar
+                        // manifest
+                        // with recommended product dependencies.
+                        if (id instanceof ProjectComponentIdentifier) {
+                            Project dependencyProject = getProject()
+                                    .getRootProject()
+                                    .project(((ProjectComponentIdentifier) id).getProjectPath());
+                            return Stream.of(
+                                    dependencyProject.getTasks().withType(ConfigureProductDependenciesTask.class));
+                        }
+                        return Stream.empty();
+                    })
+                    .forEach(emptyFileCollection::builtBy);
+            return emptyFileCollection;
+        });
+    }
+
+    @Input
+    final Property<String> getServiceName() {
+        return serviceName;
+    }
+
+    @Input
+    final Property<String> getServiceGroup() {
+        return serviceGroup;
+    }
+
+    @Input
+    final ListProperty<ProductDependency> getProductDependencies() {
+        return productDependencies;
+    }
+
+    @Input
+    final SetProperty<ProductId> getOptionalProductIds() {
+        return optionalProductIds;
+    }
+
+    @Input
+    final SetProperty<ProductId> getIgnoredProductIds() {
+        return ignoredProductIds;
+    }
+
+    @Input
+    final SetProperty<ProductId> getInRepoProductIds() {
+        return inRepoProductIds;
+    }
+
+    /**
+     * Contents of the given configuration.  Cannot list the configuration itself as an input property because the
+     * caching calculations attempt to resolve the configuration at config time.  This can lead to an error for
+     * configs that connot be resolved directly at that stage.  Caching and up-to-date calcs thus use this property.
+     */
+    @Input
+    final Set<String> getProductDependenciesConfig() {
```

Comment:
can we move this to use the artifactTransfomer pattern that we use for diagnostics?

Reply:
I see how it would be done (neat!), but I am nervous about time it will take and getting it right. All the current changes are pretty much just refactoring with the one new additional feature of being able to inject the discovered product dependencies. Can we please hold off on that improvement so that the rest of the change can move forward? I’ll circle back to it in the next couple of weeks as we are in a “soaking” period with the new features in some projects.